cmd/eksctl-anywhere/cmd/generatehardware.go

```diff
@@ -62,105 +54,90 @@ func init() {
-	if !hOpts.dryRun && hOpts.tinkerbellIp == "" {
-		return errors.New("tinkerbell-ip is required, please specify it using --tinkerbell-ip")
+	if err := validateOptions(hOpts); err != nil {
+		return err
 	}
 
-	csv, err := hardware.NewCsvParser(hOpts.csvPath)
+	csvFile, err := os.Open(hOpts.csvPath)
 	if err != nil {
-		return err
+		return fmt.Errorf("csv: %v", err)
 	}
 
-	defer csv.Close()
+	outputDir, err := hardware.CreateDefaultDir()
+	if err != nil {
+		return err
+	}
 
-	json, err := hardware.NewJsonParser(hOpts.outputPath)
+	jsonDir, err := hardware.CreateDefaultJsonDir()
 	if err != nil {
 		return err
 	}
 
-	defer json.CleanUp()
+	hardwareYaml, err := os.Create(filepath.Join(outputDir, hardware.DefaultHardwareManifestYamlFilename))
+	if err != nil {
+		return fmt.Errorf("tinkerbell manifest yaml: %v", err)
+	}
+	yamlWriter := hardware.NewTinkerbellManifestYaml(hardwareYaml)
 
-	yaml, err := hardware.NewYamlParser(hOpts.outputPath)
+	var journal hardware.Journal
+	jsonFactory, err := hardware.RecordingTinkerbellHardwareJsonFactory(jsonDir, &journal)
 	if err != nil {
 		return err
 	}
+	jsonWriter := hardware.NewTinkerbellHardwareJsonWriter(jsonFactory)
 
-	defer yaml.Close()
+	reader, err := hardware.NewCsvReader(csvFile)
+	if err != nil {
+		return fmt.Errorf("csv: %v", err)
+	}
 
-	var tink *executables.Tink
-	if !hOpts.dryRun {
+	writer := hardware.NewTeeWriterWith(yamlWriter, jsonWriter)
+	validator := hardware.NewDefaultMachineValidator()
 
-		if err := networkutils.ValidateIP(hOpts.tinkerbellIp); err != nil {
-			return fmt.Errorf("tinkerbell-ip is not valid: %v", err)
-		}
+	if err := hardware.TranslateAll(reader, writer, validator); err != nil {
+		return err
+	}
 
-		executableBuilder, close, err := executables.NewExecutableBuilder(ctx, executables.DefaultEksaImage())
+	if !hOpts.dryRun {
+		tink, close, err := tinkExecutableFromOpts(ctx, hOpts)
 		if err != nil {
-			return fmt.Errorf("unable to initialize executables: %v", err)
+			return err
 		}
-		defer close.CheckErr(ctx)
+		defer close.Close(ctx)
 
-		if !networkutils.IsPortValid(hOpts.grpcPort) {
-			return fmt.Errorf("grpc-port %s is invalid", hOpts.certPort)
+		if err := hardware.RegisterTinkerbellHardware(ctx, tink, journal); err != nil {
+			return err
 		}
+	}
 
-		if !networkutils.IsPortValid(hOpts.certPort) {
-			return fmt.Errorf("cert-port %s is invalid", hOpts.certPort)
-		}
+	return nil
+}
 
-		cert := fmt.Sprintf("http://%s:%s/cert", hOpts.tinkerbellIp, hOpts.certPort)
-		grpc := fmt.Sprintf("%s:%s", hOpts.tinkerbellIp, hOpts.grpcPort)
-		tink = executableBuilder.BuildTinkExecutable(cert, grpc)
+func validateOptions(opts *hardwareOptions) error {
+	if err := networkutils.ValidateIP(opts.tinkerbellIp); err != nil {
+		return fmt.Errorf("invalid tinkerbell-ip: %v", err)
 	}
 
-	for {
-		items, err := csv.Read()
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			return fmt.Errorf("failed reading csv: %v", err)
-		}
-
-		var id string
```

Comment:
What happens if user provides guids in csv as we do now during e2e? Are you still accounting for this? The guids provided should be used rather than generating them

Reply:
This is fixed in a non-intutive manner. TL;DR the business logic is, appropriately, wrapped up in the `CsvReader` where it allows the Csv ID column to be optional and applies a generated UUID if not present.